### DISCUSSION TITLE
검색 기능 의사코드

### DISCUSSION THREAD

**Discussion by eunohhh** [nbc-final-a10/TripBuddy#61] [아이디어]
### 옵션 1 따로 가고요.

(옵션1 고성능 검색이 가능할 경우 시도 - 자동완성, 등등)
옵션 1 - 키워드 
### 옵션 2~9 따로 갑니다.

(아래는 완전일치)
옵션 2 - 지역? - trip_destination
옵션 3 - 일시? - trip_start_date / trip_end_date
옵션 4 - 선호성별? - trip_wanted_sex 

옵션 5 - 경비? - trip_cost <<< 추후에 뺄 수도 있음
옵션 6 - 인원수? - trip_max_buddies_counts <<< 추후에 뺄 수도 있음
옵션 7 - 만남장소? - trip_meet_location <<< 추후에 뺄 수도 있음
(만약에 속도느리면 인덱싱 걸고...)

(아래2개는 세개 중 한개라도 일치하는것 다 가져온다음에, 가능하면 우선순위로 필터링)
옵션 8 - 테마? - trip_themes
옵션 9 - 선호버디즈성향? - trip_wanted_buddies

==> 1 따로 가서 배열하나 받고, 2~9 따로 가서 배열하나 받아서 중복 제거

### 옵션 1 키워드만 했을때

키워드만 검색했을때 로직은 옵션2~10까지는 고려하지 않고,
trips 테이블에서 isValidate 가 true 인 모임 중에서
trip_title 과 trip_content 에서 키워드와 일치하는 trip_id를 뽑아낸다. 
### 옵션이 하나만(2~7) 존재할 경우

trips 테이블에서 isValidate 가 true 이면서
trip_(column) 이 일치하는 trip_id 를 뽑아낸다.
### 예외) 옵션 8 / 옵션 9 의 경우

trips 테이블에서, isValidate 가 true 인 것들만남기고
user가 선택한 chip 최대 세개를 기준으로
trip_theme1, trip_theme2, trip_theme3 에서 한개라도 걸리는 trip을 다 불러온다음

```javascript
// 옵션 8, 9 에 대한 함수
// pool1, pool2, pool3 는 바꿔야 합니다
async function searchTripDocuments(searchTerms) {
  let query = supabase
    .from('trips')
    .select('*')
    .filter('isValidate', 'eq', true);

  const searchConditions = searchTerms.map(term => {
    return `(pool1 ILIKE '%${term}%' OR pool2 ILIKE '%${term}%' OR pool3 ILIKE '%${term}%')`;
  }).join(' OR ');

  query = query.or(searchConditions);

  const { data, error } = await query;

  if (error) {
    console.error('Error searching documents:', error);
  } else {
    console.log('Search results:', data);
  }
}

const searchTerms = ['one', 'two', 'three'];
searchDocuments(searchTerms);

// 역시 pool1, pool2, pool3 는 바꿔야 합니다
async function searchBuddiesDocuments(searchTerms) {
  let query = supabase
    .from('documents')
    .select('*')
    .filter('isValidate', 'eq', true);

  const searchConditions = searchTerms.map(term => {
    return `(pool1 ILIKE '%${term}%' OR pool2 ILIKE '%${term}%' OR pool3 ILIKE '%${term}%')`;
  }).join(' OR ');

  query = query.or(searchConditions);

  const { data, error } = await query;

  if (error) {
    console.error('Error searching documents:', error);
  } else {
    console.log('Search results:', data);
  }
}

const searchTerms = ['one', 'two', 'three'];
searchDocuments(searchTerms);
```

(도전과제 - 위 함수의 결과로 리턴된 배열에서 몇개가 일치하는지 퍼센트를 계산해서 그 순서대로 배열 순서 재배치)

```js
// 2~7번 까지를 결합할 수 있는 코드
// filters는 보내줘야함
// 동적으로 추가할 조건들 (클라이언트에서 보내면됨 아래를)
const filters = [
{ column: 'blabla', operator: 'eq', value: variable2 },
{ column: 'blabla2', operator: 'eq', value: variable3 },
// 필요한 만큼 추가 가능
];
  
async function searchTripDocuments(searchTerms, filters) {
  let query = supabase
    .from('trips')
    .select('*')
    .filter('isValidate', 'eq', true);

  // 각 조건을 query에 추가
  filters.forEach(filter => {
    query = query.filter(filter.column, filter.operator, filter.value);
  });

  // OR 조건을 만들어 여러 컬럼에 대해 검색어를 일치시키기
  const searchConditions = searchTerms.map(term => {
    return `(pool1 ILIKE '%${term}%' OR pool2 ILIKE '%${term}%' OR pool3 ILIKE '%${term}%')`;
  }).join(' OR ');

  query = query.or(searchConditions);

  const { data, error } = await query;

  if (error) {
    console.error('Error searching documents:', error);
  } else {
    console.log('Search results:', data);
  }
}

const searchTerms = ['one', 'two', 'three'];
const variable2 = 'some_value1';
const variable3 = 'some_value2';

searchTripDocuments(searchTerms);
```

**Comment by Captain-Kim** [comment]
## 트러블 슈팅

- 나이 제한은 최초 생각했으나 그 나이를 사용자가 가입할 때 입력하는데 그것을 검증하려면 NICE 인증 등을 해야 하고, 인증 절차가 없으면 사용자가 입력한 나이가 진짜인지 아닌지 검증할 방법이 없어 의미없는 데이터라고 판단하여 삭제함.
- jsonb에서 특정 헤시 태그를 일일이 검사하는 것이 비용이 많이 발생한다고 함 -> 컬럼으로 나눠서 하는 것이 베스트인 것 같음 -> 그러나 jsonb로 관리하기로 결정 -> 그럼 text든 boolean이든 모든 컬럼에 인덱싱을 걸어 놓으면 유리한 것 아닌가? -> 맞지만 서버 컴퓨터의 리소스를 쓰기 때문에 불필요한데 걸어 놓으면 쓰기 능력 저하
- jsonb vs 단일 컬럼 장단점
  - jsonb
    - 헤시태그 개수가 몇 개가 되든지 유연하게 늘릴 수 있음.
    - 인덱스 생성 및 유지 관리 비용이 높음. 스토리지 오버헤드가 발생할 수 있음.
  - 단일 컬럼
    - 단일 값에 대한 검색이 빠르고 jsonb보다 공간을 적게 차지함.
    - 헤시태그 컬럼이 3개면 3개 이상으로 늘리려면 컬럼을 추가해야 하는데 유지보수가 어려울 수 있음.
  - jsonb vs 단일컬럼 선택 시 고려 사항
    - 데이터 변경 가능성이 있나? 헤시태그 수가 늘어나거나 줄어들 수 있나? -> 
- 키워드 검색, 지역 검색, 여행 일시 검색 … 헤시태그 3개까지 검색이 2개… etc… 검색 옵션이 굉장히 많은 경우 -> 그 중 어떤 걸 유저가 선택할 지 모르는데, 하나의 옵션만 선택할 수도 있고 어떤 조합으로 검색할 지 예측이 안 간다 -> 모든 경우의 수를 하드코딩 해서 fetch 함수를 작성해야 하는가? -> 동적 옵션을 추가하는 방법으로 설정
```js
let query = supabase.from('meetings').select('*');

if (state1 !== null) {
  query = query.filter('a', 'eq', state1);
}
if (state2 !== null) {
  query = query.filter('b', 'eq', state2);
}
if (state3 !== null) {
  query = query.filter('c', 'eq', state3);
}

const { data, error } = await query;
```
- 그런데 조건이 너무 많아서 속도를 보장 못할 것 같다 -> trips 컬럼에서 검색과 관련된 컬럼 전부 sql 인덱싱(서버 컴퓨터 리소스 사용한다는 단점은 고려해야 함, 쓰기 속도 저하됨) -> 이것도 안 되면 데이터 캐싱
